feat: disable wallet buttons for accounts that cannot sign transactions

k-g-j opened the pull request:
## **Description**

This PR disables certain buttons in the `WalletActions` component when the selected account cannot sign transactions. This change improves the user experience by preventing attempts to perform actions that require transaction signing when the account cannot do so.

1. The reason for the change is to prevent users from attempting actions that their current account cannot perform.
2. The improvement is that Buy, Sell, Send, Swap, and Bridge buttons are now disabled when the selected account cannot sign transactions, providing clear visual feedback to the user about available actions.

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/570

## **Manual testing steps**

1. Go to the [SSK](https://metamask.github.io/snap-simple-keyring/latest/) and create an account
2. Update the account using this JSON, replacing the address and ID fields with the new account information (this update step removes `eth_signTransaction` from the account's methods)
```json
{
    "id": "your new account ID",
    "address": "you new account address",
    "options": {},
    "methods": [
        "personal_sign",
        "eth_sign",
        "eth_signTypedData_v1",
        "eth_signTypedData_v3",
        "eth_signTypedData_v4"
    ],
    "type": "eip155:eoa"
}
```
3. Switch to the account that cannot sign transactions if not already there
4. Verify that the Buy, Sell, Send, Swap, and Bridge buttons are visually disabled with a message and not clickable
5. Click on the asset details view
6. Verify that the Buy, Sell, Send, Swap, and Bridge buttons are visually disabled and not clickable
7. Switch back to an account that can sign transactions
8. Verify that the buttons are now enabled and functional in wallet view and asset details view

## **Screenshots/Recordings**

### **Before**

Wallet view
<img src="https://github.com/user-attachments/assets/d0cdcd63-cec5-4641-95a4-a47450319282" width=350 /> 

Asset details view
![simulator_screenshot_5FB98682-1B94-440A-9E12-BDE6D8B2B1C3](https://github.com/user-attachments/assets/ceac95af-3653-4178-9718-f75ff97e239f)


### **After**

Wallet view
![Simulator Screenshot - iPhone 15 Pro - 2024-10-24 at 15 43 57](https://github.com/user-attachments/assets/b3a4ed7c-3a79-413b-9ce2-6b05a5683de0)

Asset details view
![Simulator Screenshot - iPhone 15 Pro - 2024-10-24 at 15 44 06](https://github.com/user-attachments/assets/8962ef77-37c8-4bad-8d82-559dc1a808ce)


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

gantunesr commented on app/components/UI/WalletAction/WalletAction.types.ts:
```diff
@@ -1,14 +1,66 @@
+  Receive = 'Receive',
+}
+
+export interface WalletActionStrings {
+  title: string;
+  description: string;
+  disabledDescription: string;
+}
+
+export const walletActionStrings: Record<
+  WalletActionType,
+  WalletActionStrings
+> = {
```
I suggest a different naming for these types instead of using the word `string`

```suggestion
export interface WalletActionDetail {
  title: string;
  description: string;
  disabledDescription: string;
}

export const walletActionDetails: Record<
  WalletActionType,
  WalletActionDetail
> = {
```

k-g-j replied:
renamed here [20e79f9](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/20e79f9f6747da06cc98ca81fd14d49cf471c11a)

gantunesr commented on app/components/UI/WalletAction/WalletAction.types.ts:
```diff
@@ -1,14 +1,66 @@
+  WalletActionStrings
+> = {
+  [WalletActionType.Buy]: {
+    title: strings('asset_overview.buy_button'),
+    description: strings('asset_overview.buy_description'),
+    disabledDescription: strings('wallet.disabled_button.buy'),
```
Would it be better for these strings to also be inside `asset_overview` like the title and description?

k-g-j replied:
good idea! [845c66f](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/845c66f17a2a2c164d0a942319d3f7995675b0c2)

gantunesr commented on app/components/Views/WalletActions/WalletActions.test.tsx:
```diff
@@ -180,4 +187,48 @@ describe('WalletActions', () => {
 
     expect(mockNavigate).toHaveBeenCalled();
   });
+  it('should disable buttons when the account cannot sign transactions', () => {
```
nit: I know the rest of the file is using the previous convention but that shouldn't stop us of using the new one

```suggestion
  it('disables action buttons when the account cannot sign transactions', () => {
```

gantunesr reviewed (commented):
looks good overall, left some minor comments

k-g-j committed:
feat: rename 'WalletActionStrings' to 'Details'

k-g-j committed:
feat: moved disabled descriptions into asset_overview

owencraston commented on app/components/UI/WalletAction/WalletAction.types.ts:
```diff
@@ -1,14 +1,64 @@
 export interface WalletActionProps {
-  actionTitle?: string;
-  actionDescription?: string;
+  actionType?: WalletActionType;
```
why would this be optional if we are using it to derive the key information?

k-g-j replied:
It is currently optional in the implementation on the main branch [here](https://github.com/MetaMask/metamask-mobile/blob/main/app/components/UI/WalletAction/WalletAction.types.ts#L6) and making it non-optional would break [AssetDetailsActions](https://github.com/MetaMask/metamask-mobile/blob/main/app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.tsx)

owencraston replied:
true but now we are using this to derive the description which was previously a prop. I personally see no reason to make this optional. Do you?

k-g-j replied:
Agreed, we should make it non-optional 
done here [8c2b833](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/8c2b833cb918dc60e8991e3f2c8e8544aefebd9e)

owencraston commented on app/components/Views/WalletActions/WalletActions.tsx:
```diff
@@ -54,6 +56,11 @@ const WalletActions = () => {
   const { trackEvent } = useMetrics();
 
+  const selectedAccount = useSelector(selectSelectedInternalAccount);
+  const canSignTransactions = selectedAccount?.methods?.includes(
```
lets memoize this value with React.useMemo: https://react.dev/reference/react/useMemo

k-g-j replied:
smart thinking! [f5e18fb](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/f5e18fb493d7eb6311063ef787dbc0182ffa9d26)

owencraston commented on app/components/UI/WalletAction/WalletAction.types.ts:
```diff
@@ -1,14 +1,64 @@
+  disabledDescription: string;
+}
+
+export const walletActionDetails: Record<WalletActionType, WalletActionDetail> =
```
I like this a lot better.

owencraston commented on app/components/Views/WalletActions/WalletActions.test.tsx:
```diff
@@ -180,4 +187,48 @@ describe('WalletActions', () => {
   });
+  it('disables action buttons when the account cannot sign transactions', () => {
+    const mockStateWithoutSigning: DeepPartial<RootState> = {
+      ...mockInitialState,
```
Are all of these spreads necessary? Why cant you just modify the AccountsController section?

k-g-j replied:
We do need to spread the whole state unfortunately, because the buttons such as the Buy and Swap buttons need to know the context of the Chain ID and other configurations [see this test](https://github.com/MetaMask/metamask-mobile/blob/main/app/components/Views/WalletActions/WalletActions.test.tsx#L101) for an example why

owencraston commented on locales/languages/en.json:
```diff
@@ -1237,6 +1238,14 @@
     "info": "Info",
     "swap": "Swap",
     "bridge": "Bridge",
+    "disabled_button": {
```
@AlexJupiter @eriknson what do you think of this copy?

k-g-j committed:
feat: memoize 'selectedAccount' in WalletActions

k-g-j committed:
feat: memoize 'selectedAccount' in WalletActions and revert changes in ios/MetaMask.xcodeproj/project.pbxproj

Matt561 committed:
parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215624 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215619 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215615 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215606 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215597 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215562 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215557 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215549 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215524 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215487 -0400

parent fb86c8355dfc0269ed868d9a669882ae2f7254f3
author Matthew Grainger <46547583+Matt561@users.noreply.github.com> 1727214167 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215477 -0400

feat: added KeyValueRow to component-library/components-temp

parent cc011ce43e59ae70cb239408c8f0f6de10772def
author Kate Johnson <johnson.kat.g@gmail.com> 1726779250 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215830 -0400

parent cc011ce43e59ae70cb239408c8f0f6de10772def
author Kate Johnson <johnson.kat.g@gmail.com> 1726779250 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215818 -0400

parent cc011ce43e59ae70cb239408c8f0f6de10772def
author Kate Johnson <johnson.kat.g@gmail.com> 1726779250 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215815 -0400

parent cc011ce43e59ae70cb239408c8f0f6de10772def
author Kate Johnson <johnson.kat.g@gmail.com> 1726779250 -0400
committer Kate Johnson <johnson.kat.g@gmail.com> 1727215806 -0400

feat: disabled wallet overview buttons for accounts that cannot sign transactions

feat: remove 'should' from test description

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>

k-g-j committed:
revert: changes in NotificationsManager"
"

k-g-j committed:
feat: made 'actionType' for WalletAction.types non-optional

k-g-j committed:
chore: update snapshots

k-g-j committed:
feat: update AssetDetailsActions.test to use IDs for buttons

owencraston commented on app/components/UI/WalletAction/WalletAction.tsx:
```diff
@@ -8,44 +8,65 @@ import Text, {
+        </Text>
+        <Text
+          variant={TextVariant.BodyMD}
+          style={[
```
passing inline styles like this mean that on every render a new array instance will be passed as a prop. because of this the virtual dom will not memoize the props for this component. can you try and extract these out so they are not computed inline? The same should be done for the rest of the files.

k-g-j replied:
I believe I have made some improvement here in this commit [a9909ce](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/a9909ce0e9d22affaa69ed60173d8f326df456ef)

k-g-j committed:
feat: updated computed styles in AssetDetailsActions, WalletActions, and WalletAction

k-g-j committed:
feat: remove render function from AssetDetailsActions

k-g-j committed:
feat: undo changes in ios/MetaMask.xcodeproj/project.pbxproj

k-g-j commented on app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.tsx:
```diff
@@ -40,84 +40,78 @@ export const AssetDetailsActions: React.FC<AssetDetailsActionsProps> = ({
-          </Text>
-        </View>
+      {displayBuyButton &&
+        renderWalletAction(
```
removed! [88690e1](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/88690e1c50dc3996f419ac8125e25d878a85b0f9)

plasmacorral commented:
Can confirm that observation 2 above can be reproduced on a flask build of main commit 9e3d2684acfaf193cc1781680beaf7ea29f5edb8.

k-g-j committed:
feat: made 'actionType' optional prop

k-g-j committed:
chore: update snapshots

k-g-j committed:
fix: app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.test.tsx

plasmacorral commented:
Observation 1 resolved in commit 91cb7d310b775b6d3f57875c713244fa1666f8f4
<div style="display: inline-block"><img width="140" src="https://github.com/user-attachments/assets/1d132425-b98e-47bd-936b-b454f2d1e202"><img width="140" src="https://github.com/user-attachments/assets/4ca7e5c5-819d-459b-aa8f-6a0cdb3f7af5"><img width="137" src="https://github.com/user-attachments/assets/8fe34cc4-13b3-4780-9957-ee126b06f766"><img width="133" src="https://github.com/user-attachments/assets/57cd3199-1f99-4453-985e-4dce9574948c"></div>

Observation 2 is beyond the scope of this PR and was not introduced here.  Now tracked as [633](https://github.com/MetaMask/accounts-planning/issues/633).

Observation 3 can be addressed later once we are in production with the ability to disable the actions

plasmacorral commented:
Maybe only wallet view is in scope here?  These action buttons are presented and usable from the asset details pages, even for an account that does not support signing and has actions disabled from wallet view.

https://www.loom.com/share/68348eda25db4e2187a21c047a2eb32a

Is this something that should also be addressed within this PR, or do we take this incremental improvement and follow it with a subsequent PR to address asset detail pages?

k-g-j committed:
feat: disabled buttons in AssetDetails view

plasmacorral commented:
I now see that action buttons are disabled on the asset details pages. However, there is no feedback when a user clicks on these muted buttons, which can create the impression that the mobile app is unresponsive.

Although the [recording](https://www.loom.com/share/000203123b90425b92255ba96a50baf9?sid=aab86031-f120-41e2-9d82-2ea0746a3503) doesn't show it well, repeatedly clicking these muted buttons provides no feedback, which can be frustrating for users.

I'm not sure if we have an established pattern for this on mobile; it might be a good question for our DS teammates.

In the Chrome extension with Flask, the cursor changes to a barred circle, and a tooltip appears without requiring a click, explicitly stating "Not supported on this account."

While we may not be able to fully replicate the extension behavior, we have an opportunity to make the user experience more explicit on the asset details page and avoid the app feeling unresponsive.
<img width="246" alt="Screenshot 2024-10-10 at 9 37 36 PM" src="https://github.com/user-attachments/assets/05696877-609e-410a-ad15-2e88b9ffeacf">

k-g-j commented:
> I'm not sure if we have an established pattern for this on mobile; it might be a good question for our DS teammates.

cc: @eriknson and @AlexJupiter what are your thoughts?

k-g-j committed:
feat: removed string function from label rendering and removed label from account overview

k-g-j committed:
chore: strings cleanup

k-g-j committed:
Revert "chore: strings cleanup"

This reverts commit 981884aa44c3777eaaf25f15d3a3cbe78374a14d.

k-g-j committed:
Revert "feat: removed string function from label rendering and removed label from account overview"

This reverts commit 54ac028c41a0e8d82124eeb92f5a917a2e5c2d78.

k-g-j committed:
feat: fix merge conflicts with accounts controller selectors

k-g-j committed:
chore: rebase with main

plasmacorral commented:
Having an issue with local builds of flask on this branch in commit ab171e4b881505a2abdc6611a3eb397394be6854

```
ERROR  Error: createSelector expects all input-selectors to be functions, but received the following types: [undefined, function memoized(), function memoized()], js engine: hermes
 ERROR  TypeError: Cannot read property 'store' of undefined, js engine: hermes
 ERROR  Error: createSelector expects all input-selectors to be functions, but received the following types: [undefined, function memoized(), function memoized()], js engine: hermes
 ERROR  TypeError: Cannot read property 'store' of undefined, js engine: hermes
```
![Screenshot_20241018-174745](https://github.com/user-attachments/assets/8a9d3603-6e93-46fd-90c5-d97e4cd355cb)

k-g-j committed:
Merge branch 'main' into feat/disable-signing-buttons

brianacnguyen commented on app/components/UI/WalletAction/WalletAction.tsx:
```diff
@@ -1,51 +1,81 @@
+    ],
+    [styles, disabled, colors],
+  );
+
```
These style changes are not needed (and also does not follow our disabled pattern. Instead, you can just update the container style to have  `opacity: 0.5`

k-g-j replied:
okay! thanks for letting me me know :) 
I adjusted the styles to only use `opacity: 0.5` in this commit [153c801](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/153c801904870d58708663be384383abbbd098f0)

brianacnguyen reviewed (changes requested):
1 small request

k-g-j committed:
Merge branch 'main' into feat/disable-signing-buttons

k-g-j committed:
feat: updated disabled styles to use opacity

k-g-j committed:
Merge branch 'main' into feat/disable-signing-buttons

owencraston commented on app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.tsx:
```diff
@@ -39,47 +41,51 @@ export const AssetDetailsActions: React.FC<AssetDetailsActionsProps> = ({
             onPress={onBuy}
-            iconStyle={styles.icon}
-            containerStyle={styles.containerStyle}
+            {...walletActionProps}
```
I THINK the spread operator here defeats the purpose of the memoization since spreading the object will create a new reference anyway. [See this stackoverflow thread](https://stackoverflow.com/questions/52959059/does-spread-operator-return-new-object-after-operation). I am not certain about this so please correct me if I am wrong.

owencraston replied:
the same goes for the rest of the file.

k-g-j replied:
I THINK you are right

I updated the code in this commit [37077bf](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/37077bfc24ac5a8022b1e5d05f833d3ec17bff89) so we're not creating new object references unnecessarily, which was defeating the purpose of memoization. It also makes the code more explicit about which props are being passed to each `WalletAction` component.

owencraston commented on app/components/Views/WalletActions/WalletActions.tsx:
```diff
@@ -67,132 +77,149 @@ const WalletActions = () => {
-            iconStyle={styles.icon}
             actionID={WalletActionsModalSelectorsIDs.BUY_BUTTON}
+            iconStyle={styles.icon}
+            {...walletActionBaseProps}
```
same question about spreading the memoized props.

k-g-j replied:
Same logic as your above comment and my point about creating new object references unnecessarily 
addressed here [37077bf](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/37077bfc24ac5a8022b1e5d05f833d3ec17bff89)

owencraston reviewed (commented):
Looking very good and very nice 😁 ! The code looks good I just had a few comments/questions about all the memoization of props.

owencraston commented:
you will need to run `yarn jest -u pp/components/UI/AssetOverview/AssetOverview.test.tsx` to update the snapshot and get your unit tests passing.

k-g-j committed:
feat: created touchable styles obj

k-g-j commented on app/components/UI/WalletAction/WalletAction.tsx:
```diff
@@ -1,46 +1,59 @@
   return (
     <TouchableOpacity
-      style={{ ...styles.base, ...containerStyle }}
+      style={[styles.base, containerStyle, disabled && styles.disabled]}
```
[5946bfb](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/5946bfb50a54bf9003b4ef85e56d2c47227203c4)

k-g-j committed:
feat: remove walletActionProps

k-g-j committed:
feat: remove walletActionBaseProps

k-g-j committed:
chore: update snapshots

k-g-j committed:
Merge branch 'main' into feat/disable-signing-buttons

# Conflicts:
#	app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.test.tsx
#	app/components/Views/AssetDetails/AssetDetailsActions/AssetDetailsActions.tsx

plasmacorral commented:
Tested on Pixel 5a with Android 14 with commit [f714dd1d56a1b22e79b6b53cc8a5668294a22eca](https://github.com/MetaMask/metamask-mobile/pull/11330/commits/f714dd1d56a1b22e79b6b53cc8a5668294a22eca). Buttons were enabled in the wallet view bottom bar and asset view when signing was enabled on the snap account, and disabled when signing permission was removed. Re-enabling signing restored the buttons. Checked on Eth mainnet, Sepolia, Base, and Arbitrum. No impact on button display or function for HD, imported, QR, and Ledger accounts.

brianacnguyen commented on app/components/UI/WalletAction/WalletAction.tsx:
```diff
@@ -1,46 +1,65 @@
+  const avatarStyle = useMemo(
+    () => ({
+      ...iconStyle,
+      backgroundColor: disabled ? colors.primary.muted : colors.primary.default,
```
this disabled change is not needed since the opacity would take care of color change

brianacnguyen reviewed (changes requested):
1 small request regarding avatar disabled

owencraston commented:
Do to difficulty rebasing this PR I am continuing the work over [here](https://github.com/MetaMask/metamask-mobile/pull/12145).
